src/sync/sync_session.cpp

```diff
@@ -608,28 +643,35 @@ bool SyncSession::can_wait_for_network_completion() const
     return m_state == &State::active || m_state == &State::dying;
 }
 
+bool SyncSession::can_queue_for_network_completion() const
+{
+    return m_state == &State::waiting_for_access_token || m_state == &State::inactive;
+}
+
 bool SyncSession::wait_for_upload_completion(std::function<void(std::error_code)> callback)
 {
     std::unique_lock<std::mutex> lock(m_state_mutex);
-    // FIXME: instead of dropping the callback if we haven't yet `bind()`ed,
-    // save it and register it when the session `bind()`s.
     if (can_wait_for_network_completion()) {
```

Comment:
Rather than switching on the session state, can we instead just delegate to a function on the state?